File: backend/app/routes/tasks.py
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status

from app.auth import require_auth
from app.schemas import (
    ActivityListResponse,
    TaskCreate,
    TaskListResponse,
    TaskResponse,
    TaskStatsResponse,
    TaskUpdate,
)
from app.store import TaskNotFoundError, task_store


router = APIRouter(prefix="/tasks", tags=["tasks"], dependencies=[Depends(require_auth)])


def _not_found() -> HTTPException:
    """Normalize missing task errors across task routes."""

    return HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail="Task not found",
    )


@router.get("/stats", response_model=TaskStatsResponse)
def get_task_stats() -> TaskStatsResponse:
    """Return aggregate counts used by the frontend stats panel."""

    return TaskStatsResponse(**task_store.get_stats())


@router.get("", response_model=TaskListResponse)
def list_tasks(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=10, ge=1, le=50),
    status_filter: Literal["all", "pending", "in_progress", "completed"] = Query(
        default="all",
        alias="status",
    ),
    search: str | None = Query(default=None, min_length=1, max_length=160),
    tag: str | None = Query(default=None, min_length=1, max_length=200),
    sort_by: Literal["created_at", "target_date", "status", "priority"] = "created_at",
    sort_order: Literal["asc", "desc"] = "desc",
) -> TaskListResponse:
    """List tasks with pagination, all-tags filtering, search, and sorting."""

    tag_filters = [value.strip() for value in (tag or "").split(",") if value.strip()]

    items, total, total_pages = task_store.list_tasks(
        page=page,
        page_size=page_size,
        status_filter=status_filter,
        search=search,
        tags=tag_filters,
        sort_by=sort_by,
        sort_order=sort_order,
    )
    return TaskListResponse(
        items=items,
        page=page,
        page_size=page_size,
        total=total,
        total_pages=total_pages,
    )


@router.post("", response_model=TaskResponse, status_code=status.HTTP_201_CREATED)
def create_task(payload: TaskCreate) -> TaskResponse:
    """Create a new pending task."""

    task = task_store.create_task(
        payload.title,
        priority=payload.priority,
        tags=payload.tags,
        target_date=payload.target_date,
    )
    return TaskResponse.model_validate(task)


@router.get("/{task_id}", response_model=TaskResponse)
def get_task(task_id: str) -> TaskResponse:
    """Return detail for one task."""

    try:
        return TaskResponse.model_validate(task_store.get_task(task_id))
    except TaskNotFoundError as exc:
        raise _not_found() from exc


@router.patch("/{task_id}", response_model=TaskResponse)
def update_task(task_id: str, payload: TaskUpdate) -> TaskResponse:
    """Update editable task fields."""

    try:
        return TaskResponse.model_validate(
            task_store.update_task(
                task_id,
                title=payload.title,
                status=payload.status,
                priority=payload.priority,
                tags=payload.tags,
                target_date=payload.target_date,
            )
        )
    except TaskNotFoundError as exc:
        raise _not_found() from exc


@router.put("/{task_id}/complete", response_model=TaskResponse)
def complete_task(task_id: str) -> TaskResponse:
    """Mark a task as completed."""

    try:
        return TaskResponse.model_validate(task_store.complete_task(task_id))
    except TaskNotFoundError as exc:
        raise _not_found() from exc


@router.put("/{task_id}/toggle", response_model=TaskResponse)
def toggle_task(task_id: str) -> TaskResponse:
    """Toggle completion state for a smoother frontend interaction."""

    try:
        return TaskResponse.model_validate(task_store.toggle_task(task_id))
    except TaskNotFoundError as exc:
        raise _not_found() from exc


@router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_task(task_id: str) -> Response:
    """Delete a task."""

    try:
        task_store.delete_task(task_id)
    except TaskNotFoundError as exc:
        raise _not_found() from exc
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/{task_id}/activity", response_model=ActivityListResponse)
def get_task_activity(task_id: str) -> ActivityListResponse:
    """Return the activity timeline for one task."""

    try:
        return ActivityListResponse(items=task_store.get_activity(task_id))
    except TaskNotFoundError as exc:
        raise _not_found() from exc

File: frontend/src/api/tasks.ts
import { apiRequest } from "./client";
import type {
  ActivityListResponse,
  Task,
  TaskCreateInput,
  TaskListResponse,
  TaskSortBy,
  TaskStats,
  TaskStatusFilter,
  TaskUpdateInput,
  SortOrder,
} from "./types";

export type ListTasksParams = {
  page: number;
  pageSize: number;
  status: TaskStatusFilter;
  search?: string;
  tag?: string;
  sortBy: TaskSortBy;
  sortOrder: SortOrder;
};

/**
 * Fetch a backend-filtered and sorted page of tasks.
 *
 * `tag` can contain a comma-separated list. The backend returns only tasks
 * containing every requested tag.
 */
export function listTasks(
  token: string,
  params: ListTasksParams,
): Promise<TaskListResponse> {
  // The backend owns pagination/filtering so the frontend can stay lightweight.
  const searchParams = new URLSearchParams({
    page: String(params.page),
    page_size: String(params.pageSize),
    status: params.status,
    sort_by: params.sortBy,
    sort_order: params.sortOrder,
  });

  if (params.search?.trim()) {
    searchParams.set("search", params.search.trim());
  }
  if (params.tag?.trim()) {
    searchParams.set("tag", params.tag.trim());
  }

  return apiRequest<TaskListResponse>(`/tasks?${searchParams.toString()}`, { token });
}

/** Create a task from the toolbar form payload. */
export function createTask(token: string, input: TaskCreateInput): Promise<Task> {
  return apiRequest<Task>("/tasks", {
    token,
    method: "POST",
    body: input,
  });
}

/** Fetch one task for the detail page. */
export function getTask(token: string, taskId: string): Promise<Task> {
  return apiRequest<Task>(`/tasks/${taskId}`, { token });
}

/** Convenience wrapper for the original title-only edit behavior. */
export function updateTaskTitle(
  token: string,
  taskId: string,
  title: string,
): Promise<Task> {
  return updateTask(token, taskId, { title });
}

/** Apply a partial task update, such as status, priority, tags, or due date. */
export function updateTask(
  token: string,
  taskId: string,
  input: TaskUpdateInput,
): Promise<Task> {
  return apiRequest<Task>(`/tasks/${taskId}`, {
    token,
    method: "PATCH",
    body: input,
  });
}

/** Preserve the required take-home endpoint for marking a task completed. */
export function completeTask(token: string, taskId: string): Promise<Task> {
  return apiRequest<Task>(`/tasks/${taskId}/complete`, {
    token,
    method: "PUT",
  });
}

/** Flip between completed and pending for the detail page action button. */
export function toggleTask(token: string, taskId: string): Promise<Task> {
  return apiRequest<Task>(`/tasks/${taskId}/toggle`, {
    token,
    method: "PUT",
  });
}

/** Delete a task and resolve after the backend returns 204. */
export function deleteTask(token: string, taskId: string): Promise<void> {
  return apiRequest<void>(`/tasks/${taskId}`, {
    token,
    method: "DELETE",
  });
}

/** Load aggregate counts for the task dashboard cards. */
export function getTaskStats(token: string): Promise<TaskStats> {
  return apiRequest<TaskStats>("/tasks/stats", { token });
}

/** Load the audit timeline shown on the detail page. */
export function getTaskActivity(
  token: string,
  taskId: string,
): Promise<ActivityListResponse> {
  return apiRequest<ActivityListResponse>(`/tasks/${taskId}/activity`, { token });
}
